libindy/src/services/pool/catchup.rs

```diff
@@ -115,40 +118,67 @@ impl CatchupHandler {
         if self.pending_catchup.is_some() {
             return Ok(CatchupProgress::InProgress);
         }
-        let mut votes: HashMap<(String, usize), usize> = HashMap::new();
-        for node_vote in &self.nodes_votes {
-            if let &Some(ref node_vote) = node_vote {
-                let cnt = *votes.get(&node_vote).unwrap_or(&0) + 1;
-                votes.insert((node_vote.0.clone(), node_vote.1), cnt);
-            }
-        }
+
+        let votes: HashMap<(String, usize, Option<Vec<String>>), usize> = self.nodes_votes.iter().cloned()
+            .filter_map(|e| e)
+            .fold(HashMap::new(), |mut acc, vote| {
+                let cnt = acc.get(&vote).unwrap_or(&0) + 1;
+                acc.insert(vote, cnt);
+                acc
+            });
+
         if let Some((most_popular_vote, votes_cnt)) = votes.iter().max_by_key(|entry| entry.1) {
             if *votes_cnt == self.nodes.len() - self.f {
-                let &(ref target_mt_root, target_mt_size) = most_popular_vote;
-                let cur_mt_size = self.merkle_tree.count();
-                let cur_mt_hash = self.merkle_tree.root_hash().to_base58();
-                if target_mt_size == cur_mt_size {
-                    if cur_mt_hash.eq(target_mt_root) {
-                        return Ok(CatchupProgress::NotNeeded);
-                    } else {
-                        return Err(PoolError::CommonError(CommonError::InvalidState(
-                            "Ledger merkle tree doesn't acceptable for current tree.".to_string())));
+                return match self._try_to_catch_up(most_popular_vote) {
+                    //WARNING: If we receive invalid cache then it means that cache was cleared. We need to rebuild merkle tree and try to catch up once again
+                    Err(PoolError::InvalidCacheCleared) => {
+                        self.merkle_tree = PoolWorker::restore_merkle_tree_from_pool_name(&self.pool_name)?;
+                        self._try_to_catch_up(most_popular_vote)
                     }
-                } else if target_mt_size > cur_mt_size {
-                    self.target_mt_size = target_mt_size;
-                    self.target_mt_root = target_mt_root.from_base58().map_err(|_|
-                        CommonError::InvalidStructure(
-                            "Can't parse target MerkleTree hash from nodes responses".to_string()))?;
-                    return Ok(CatchupProgress::ShouldBeStarted);
-                } else {
-                    return Err(PoolError::CommonError(CommonError::InvalidState(
-                        "Local merkle tree greater than mt from ledger".to_string())));
+                    a => a
                 }
             }
         }
         Ok(CatchupProgress::InProgress)
     }
 
+    fn _try_to_catch_up(&mut self, most_popular_vote: &(String, usize, Option<Vec<String>>)) -> Result<CatchupProgress, PoolError> {
```

Comment:
I suggest to rename `most_popular_vote` to something like `confirmed_vote_for_ledger_status`. Not sure about exact name, but here it's not just most popular, now it's confirmed by consensus.